packages/application-shell/src/components/navbar/navbar.tsx

```diff
@@ -125,58 +125,87 @@ export const ApplicationMenu = (props: ApplicationMenuProps) => {
-    subRefBoundingClientRect ?? {};
-
-  /* We want to track the left, top, width, and height of the safe area */
-  const submenuSafeAreaRefBoundingClientRect =
-    submenuSafeAreaRef.current?.getBoundingClientRect();
-  const safeAreaLeftPos = submenuSafeAreaRefBoundingClientRect?.left || 0;
-  const safeAreaTopPos = submenuSafeAreaRefBoundingClientRect?.top || 0;
-  const safeAreaWidth = submenuSafeAreaRefBoundingClientRect?.width || 0;
-  const safeAreaHeight = submenuSafeAreaRefBoundingClientRect?.height || 0;
+  const hasSubmenu =
+    Array.isArray(props.menu.submenu) && props.menu.submenu.length > 0;
+
+  const menuItemIdentifier = snakeCase(props.menu.key);
 
   const calculateSafeAreaStartPositon = useCallback(
     (e: MouseEvent) => {
-      const localX = e.clientX - safeAreaLeftPos;
-      const localY = e.clientY - safeAreaTopPos;
+      const currentSafeAreaRect =
+        submenuSafeAreaRef.current?.getBoundingClientRect();
+
+      const menuItemRect = document
+        .querySelector(`[data-menuitem="${menuItemIdentifier}"]`)
+        ?.getBoundingClientRect();
+
+      const currentSafeAreaLeft = currentSafeAreaRect?.left || 0;
+      const currentSafeAreaTop = currentSafeAreaRect?.top || 0;
+      const currentSafeAreaWidth = currentSafeAreaRect?.width || 0;
+      const currentSafeAreaHeight = currentSafeAreaRect?.height || 0;
+
+      const localX = e.clientX - currentSafeAreaLeft;
+      const localY = e.clientY - currentSafeAreaTop;
+
+      let minXPercent = 0;
+      if (menuItemRect && currentSafeAreaRect && currentSafeAreaWidth > 0) {
+        const minX = Math.max(menuItemRect.left - currentSafeAreaRect.left, 0);
+        minXPercent = (minX / currentSafeAreaWidth) * 100;
+      } else if (
+        menuItemRect &&
+        currentSafeAreaRect &&
+        currentSafeAreaWidth === 0
+      ) {
+        // If safe area has no width but menu item exists, consider minXPercent to be 0
+        minXPercent = 0;
+      }
 
-      setPercentageX((localX / safeAreaWidth) * 100);
-      setPercentageY((localY / safeAreaHeight) * 100);
+      // Calculate nextX, ensuring not to divide by zero
+      let nextX = minXPercent;
+      if (currentSafeAreaWidth > 0) {
+        nextX = (localX / currentSafeAreaWidth) * 100;
+      }
+
+      // Calculate nextY, ensuring not to divide by zero
+      let nextY = 0;
+      if (currentSafeAreaHeight > 0) {
+        nextY = (localY / currentSafeAreaHeight) * 100;
+      }
+
+      // Clamp X to menu item's left border (minXPercent) and safe area bounds (100%)
+      nextX = Math.max(nextX, minXPercent);
+      nextX = Math.min(nextX, 100);
+      if (!Number.isFinite(nextX)) {
+        nextX = minXPercent;
+      }
+
+      // Clamp Y to safe area bounds (0% to 100%)
+      nextY = Math.min(Math.max(nextY, 0), 100);
+      if (!Number.isFinite(nextY)) {
+        nextY = 0;
+      }
+
+      setPercentageX(nextX);
+      setPercentageY(nextY);
```

Comment:
This is to ensure we don't end up with values above 100% and evaluations like `Infinity % Infinity %` in case of division by 0